### PR TITLE
README: describe how to use with external runners and builtin generators

### PR DESCRIPTION
**Pull request by ligurio**
- describe usage with external test-runners (as followup for #29)
- describe builtin generators (I believe it would be useful to have these descriptions in README. Yes, I know about documentation generated by `ldoc`.)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,8 +33,11 @@ Some example properties can be found [here](https://github.com/luc-tielen/lua-qu
 
 ## Usage
 
-Lua Quickcheck exposes a test-runner as a command-line interface for running
-all property based tests in your codebase. Assuming `lqc` is installed and
+There are two ways to use Lua QuickCheck library: using its own test-runner
+`lqc` or using external test-runner.
+
+Lua QuickCheck exposes a test-runner as a command-line interface for running
```
Comment (luc-tielen) [inline README.md:41]:
```suggestion 
### Using the LQC test runner

Lua QuickCheck exposes a test-runner as a command-line interface for running
```

File: README.md [hunk 2]
```diff
@@ -46,6 +49,96 @@ By default, `lqc` will look for all `*.lua` and `*.moon` files in the current
 directory for .lua and .moon files and execute all of them; but it is also
 possible to specify a specific set of files or directories to execute.
 
+In more or less mature projects tests already exist and run using test-runner
```
Comment (luc-tielen) [inline README.md]:
```suggestion
### Using an external test runner

In mature projects tests already exist and run using test-runner
```

File: README.md [hunk 3]
```diff
@@ -46,6 +49,96 @@ By default, `lqc` will look for all `*.lua` and `*.moon` files in the current
...
+provided by test framework. For example `busted`, `luatest` have their own
+test-runners.  In such cases it is still possible to use Lua QuickCheck,
+everything to need is import required modules and run `lqc.check()` in
+testcase.
```
Comment (luc-tielen) [inline README.md]:
```suggestion
test-runners.  In such cases it is still possible to use Lua QuickCheck.
All you need to do is import required modules and run `lqc.check()` in
each testcase.
```

File: README.md [hunk 4]
```diff
@@ -46,6 +49,96 @@ By default, `lqc` will look for all `*.lua` and `*.moon` files in the current
...
+
+Below is example of test `test/sum_test.lua` to use with `luatest` library
```
Comment (luc-tielen) [inline README.md]:
```suggestion
Below is an example of test `test/sum_test.lua` for use with the `luatest` library:
```

File: README.md [hunk 5]
```diff
@@ -46,6 +49,96 @@ By default, `lqc` will look for all `*.lua` and `*.moon` files in the current
...
+```
+
+To run test execute `luatest ./test/test_sum.lua`:
+```
+$ luatest ./test/sum_test.lua
+.
+Ran 1 tests in 0.017 seconds, 1 success, 0 failures
+```
+
+And similar with test library `busted`:
```
Comment (luc-tielen) [inline README.md]:
```suggestion
And here's an example with test library `busted`:
```

File: README.md [hunk 6]
```diff
@@ -46,6 +49,96 @@ By default, `lqc` will look for all `*.lua` and `*.moon` files in the current
...
+      check = function(n)
+        return sum_up_to(n) == (n + 1) * n / 2
+      end
+    }
+    lqc.check()
+  end)
+end)
+```
+
+To run test execute: `busted spec/sum_spec.lua`
```
Comment (luc-tielen) [inline README.md]:
```suggestion
To run the tests, execute the following command: `busted spec/sum_spec.lua`
```

**Review by luc-tielen — CHANGES_REQUESTED** [review]
Left some minor suggestions, looks good for the rest.

Thanks!

**Comment by ligurio** [comment]
> Left some minor suggestions, looks good for the rest.

Thanks for suggestions! Everything committed, squashed (also I have added `Co-authored-by` to mark you as co-author) and force-pushed.
Seems change with descriptions for generators requires review too - I'm stuck with description for some of them.

**Comment by ligurio** [comment]
Unfortunately CI is broken, sorry for that.
I'll fix it in a separate commit in the same PR (if you have no objections).
right now it can start, but:
- Lua == 5.1 failed due to segmentation fault in busted (https://github.com/luc-tielen/lua-quickcheck/pull/30/checks?check_run_id=2333602586)
- Lua 5.2 failed with strange reason:
```
Error: Failed installing dependency: https://luarocks.org/luassert-1.8.0-0.rockspec - Error fetching file: Failed downloading https://luarocks.org/luassert-1.8.0-0.rockspec - failed downloading https://luarocks.org/luassert-1.8.0-0.rockspec
say 1.3-1 is now installed in /home/runner/work/lua-quickcheck/lua-quickcheck/lua_install (license: MIT <http://opensource.org/licenses/MIT>)
```
although file `https://luarocks.org/luassert-1.8.0-0.rockspec` can be successfully downloaded on my personal machine.
- LuaJIT jobs failed to upload code coverage data:
```
Uploading ...
Upload error: Couldn't find a repository matching this job.
Raw data:	Couldn't find a repository matching this job.
--------------------
Upload error: Couldn't find a repository matching this job.
Raw data:	Couldn't find a repository matching this job.
--------------------
```

**Comment by luc-tielen** [comment]
@ligurio Yeah, it never is that easy to switch over CI.. I'd look into the segfault first. If we can't fix that; we might have to go back to old way of CI? :thinking:

**Comment by ligurio** [comment]
> Yeah, it never is that easy to switch over CI.. I'd look into the segfault first. If we can't fix that; we might have to go back to old way of CI?

I suspect the root cause is not in CI itself but in actions described by me. CI workflow runs on Ubuntu 20.04, I have the same OS on my laptop and everything works fine there. I'll try to investigate problem more and try to fix it.

**Comment by ligurio** [comment]
I have found a [ticket](https://github.com/moteus/luacov-coveralls) about random luajit segfaults on running busted and also related [conversation](https://www.freelists.org/post/luajit/random-segfaults-when-running-tests-with-busted) in luajit mailing list.
Someone suggested a [solution](https://github.com/Olivine-Labs/luassert/pull/150) with small Lua snippet. I have applied it to lua-quickcheck and now CI is red but due to failed steps with upload code coverage data to coveralls service. I suspect `lua-coveralls` requires environment variable with coveralls token (see https://github.com/moteus/luacov-coveralls).

For me this solution looks like a dirty hack but I don't know better solution.

**Comment by luc-tielen** [comment]
@ligurio maybe we should split this PR into multiple parts? Just README docs in this 1, fixing CI in another? CI is broken anyway, so I will merge in when docs are updated.

**Comment by ligurio** [comment]
Agree. I have failed to fix CI quickly so it's better to split PRs. I'll do.

**Comment by ligurio** [comment]
@luc-tielen done

**Review by luc-tielen — APPROVED** [review]

**Pull request merged**